phoenix-core-client/src/main/java/org/apache/phoenix/expression/function/BsonConditionExpressionFunction.java

```diff
@@ -0,0 +1,137 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.phoenix.expression.function;
+
+import java.util.List;
+
+import org.apache.phoenix.expression.util.bson.DocumentComparisonExpressionUtils;
+import org.bson.BsonDocument;
+import org.bson.BsonString;
+import org.bson.BsonValue;
+import org.bson.RawBsonDocument;
+
+import org.apache.hadoop.hbase.io.ImmutableBytesWritable;
+import org.apache.phoenix.expression.Expression;
+import org.apache.phoenix.expression.util.bson.SQLComparisonExpressionUtils;
+import org.apache.phoenix.parse.FunctionParseNode;
+import org.apache.phoenix.parse.BsonConditionExpressionParseNode;
+import org.apache.phoenix.schema.tuple.Tuple;
+import org.apache.phoenix.schema.types.PBoolean;
+import org.apache.phoenix.schema.types.PDataType;
+import org.apache.phoenix.schema.types.PBson;
+import org.apache.phoenix.schema.types.PVarbinary;
+import org.apache.phoenix.schema.types.PVarchar;
+import org.apache.phoenix.thirdparty.com.google.common.base.Preconditions;
+
+@FunctionParseNode.BuiltInFunction(name = BsonConditionExpressionFunction.NAME,
+    nodeClass = BsonConditionExpressionParseNode.class,
+    args =
+        {
+            @FunctionParseNode.Argument(allowedTypes = {PBson.class, PVarbinary.class}),
+            @FunctionParseNode.Argument(allowedTypes = {PBson.class, PVarbinary.class},
+                    isConstant = true)
+        })
+public class BsonConditionExpressionFunction extends ScalarFunction {
+
+    public static final String NAME = "BSON_CONDITION_EXPRESSION";
+
+    public BsonConditionExpressionFunction() {
+    }
+
+    public BsonConditionExpressionFunction(List<Expression> children) {
+        super(children);
+        Preconditions.checkNotNull(getConditionExpression());
+    }
+
+    @Override
+    public String getName() {
+        return NAME;
+    }
+
+    @Override
+    public boolean evaluate(Tuple tuple, ImmutableBytesWritable ptr) {
+        if (!getColValExpr().evaluate(tuple, ptr)) {
+            return false;
+        }
+        if (ptr == null || ptr.getLength() == 0) {
+            return false;
+        }
+
+        RawBsonDocument rawBsonDocument =
+            (RawBsonDocument) PBson.INSTANCE.toObject(ptr, getColValExpr().getSortOrder());
+
+        if (!getConditionExpression().evaluate(tuple, ptr)) {
+            return false;
+        }
+        if (ptr.getLength() == 0) {
+            return false;
+        }
+
+        final RawBsonDocument conditionExpressionBsonDoc;
+        if (getConditionExpression().getDataType() == PVarchar.INSTANCE) {
+            String conditionExpression =
+                    (String) PVarchar.INSTANCE.toObject(ptr,
+                            getConditionExpression().getSortOrder());
+            if (conditionExpression == null || conditionExpression.isEmpty()) {
+                return true;
```

Comment:
does true here imply returning the complete bson? mean like no condition so return all?

Reply:
since condition is empty, it is true so it is meant to be evaluated to true

Reply:
right but the ptr which contains the result will be currently set as the total bson document right?

Reply:
Correct